doc/cfconcat.1
'\" -*- coding: us-ascii -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH "cfconcat " 1 "12 May 2015" 0.3.2 "User Commands"
.SH NAME
cfconcat
\- Access groups of files as if they were a single large file 
.SH SYNOPSIS
'nh
.fi
.ad l
\fBcfconcat\fR \kx
.if (\nx>(\n(.l/2)) .nr x (\n(.l/5)
'in \n(.iu+\nxu
[OPTION]\&... [-h] [FILES]
'in \n(.iu-\nxu
.ad b
'hy
.SH DESCRIPTION
\fBcfconcat\fR is a wrapper around
\fBconcat-fuse\fR that allows accessing groups of
files, selected by a glob pattern, as if they were a single
large file. It works by mounting the virtual filesystem
\fBconcat-fuse\fR and then accessing a file within
it that acts as if it contained all the individual files
concatenated.
.PP
\fBcfconcat\fR is especially useful for doing gapless
playback of video files (when the video format allows it) or for
accessing archive files that have been `split` into multiple
parts and where a temporary file would be to slow, large and
cumbersome to deal with.
.SH OPTIONS
.TP 
\*(T<\fB\-h\fR\*(T>, \*(T<\fB\-\-help\fR\*(T>
Display help text and exit.
.TP 
\*(T<\fB\-u\fR\*(T>, \*(T<\fB\-\-unmount\fR\*(T>
Unmount concat-fuse
.TP 
\*(T<\fB\-e\fR\*(T> \fIEXE\fR, \*(T<\fB\-\-exe\fR\*(T> \fIEXE\fR 
Use EXE instead of concat-fuse
.TP 
\*(T<\fB\-m\fR\*(T> \fIMOUNTPOINT\fR, \*(T<\fB\-\-mountpoint\fR\*(T> \fIMOUNTPOINT\fR
Use MOUNTPOINT instead of default
.TP 
\*(T<\fB\-n\fR\*(T>, \*(T<\fB\-\-dry\-run\fR\*(T>
Don't send file list to concat-fuse, print to stdout
.TP 
\*(T<\fB\-k\fR\*(T>, \*(T<\fB\-\-keep\fR\*(T>
Continue even when files are missing
.TP 
\*(T<\fB\-\-ext\fR\*(T> \fISUFFIX\fR
Add an suffix to the virtual filename, useful in cases
where an application can't handle arbitary filenames and
require an extension for proper detection. The
\fISUFFIX\fR is appended directly to
the filename and should thus include a dot (i.e.
\*(T<\fI\&.zip\fR\*(T> and not just
\*(T<\fIzip\fR\*(T>
.TP 
\fIFILES\fR
Files to virtually concatenate
.TP 
\*(T<\fB\-\-from\-file\fR\*(T> \fIFILE\fR
Read files from FILE, newline separated
.TP 
\*(T<\fB\-\-from\-file0\fR\*(T> \fIFILE\fR
Read files from FILE, \*(T<\e0\*(T> separated
.TP 
\*(T<\fB\-\fR\*(T>
Read files from stdin
.TP 
\*(T<\fB\-0\fR\*(T>
Read files from stdin, '\e0' separated
.TP 
\*(T<\fB\-g\fR\*(T> \fIGLOB\fR, \*(T<\fB\-\-glob\fR\*(T> \fIGLOB\fR
Use glob the given \fIGLOB\fR
.TP 
\*(T<\fB\-z\fR\*(T> \fIARCHIVE\fR, \*(T<\fB\-\-zip\fR\*(T> \fIARCHIVE\fR
Read files from a .zip archive instead of from the filesystem
.SH EXAMPLE
A simple example of \fBcfconcat\fR use would look like this:
.PP
.nf
\*(T<$ echo "File Number One" > /tmp/1.txt
$ echo "File Number Two" > /tmp/2.txt
$ cat $(cfconcat /tmp/*.txt)
File Number One
File Number Two\*(T>
.fi
.PP
The glob pattern can also provided to \fBcfconcat\fR
directly instead of being expanded by the Shell, this allows
circumventing:
.PP
.nf
\*(T<bash: /home/ingo/bin/cfconcat: Argument list too long\*(T>
.fi
.PP
Furthermore \fBtouch\fR'ing the virtual file or
resubmitting the glob pattern will update the virtual file in
case new files have arrived matching the pattern. Note that the
glob pattern needs to be quoted when supplied to
\fB--glob\fR.
.PP
.nf
\*(T<$ echo "File Number One" > /tmp/1.txt
$ echo "File Number Two" > /tmp/2.txt
$ cat $(cfconcat \-\-glob "/tmp/*.txt")
File Number One
File Number Two\*(T>
.fi
.PP
Once done with the virtual file the virtual filesystem can be
unmounted with:
.PP
.nf
\*(T<$ cfconcat \-u\*(T>
.fi
.SH FILES
.TP 
\*(T<\fI$XDG_RUNTIME_DIR/concats\-fuse/\fR\*(T>, \*(T<\fI$HOME/.concat\-fuse/\fR\*(T>
Default mount location that \fBcfconcat\fR uses
for \fBconcat-fuse\fR
.SH "SEE ALSO"
\fBconcat-fuse\fR(1),
\fBfuse\fR(8)


cfconcat.py
#!/usr/bin/env python

# concat-fuse
# Copyright (C) 2015 Ingo Ruhnke <grumbel@gmx.de>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from __future__ import print_function

import argparse
import glob
import hashlib
import os
import subprocess
import sys
import tempfile


class ConcatFuse:

    @staticmethod
    def unmount(basedir=None):
        if basedir is None:
            basedir = ConcatFuse.find_basedir()

        if os.path.ismount(basedir):
            subprocess.check_call(["fusermount", "-u", basedir])

    @staticmethod
    def find_basedir():
        runtime_base_dir = os.environ.get("XDG_RUNTIME_DIR")
        if runtime_base_dir:
            basedir = os.path.join(runtime_base_dir, "concat-fuse")
        else:
            basedir = os.path.join(os.environ["HOME"], ".concat-fuse")
        return basedir

    def __init__(self, basedir=None, concat_fuse_exe=None):
        """Mounts concat-fuse to basedir when needed"""

        if basedir is None:
            self.basedir = ConcatFuse.find_basedir()
        else:
            self.basedir = basedir

        if concat_fuse_exe is None:
            self.concat_fuse_exe = "concat-fuse"
        else:
            self.concat_fuse_exe = concat_fuse_exe

        if not os.path.exists(self.basedir):
            os.mkdir(self.basedir)

        if not os.path.ismount(self.basedir):
            subprocess.check_call([self.concat_fuse_exe, self.basedir])

        self.check_version()

    def check_version(self):
        expected_version = "3"
        with open(os.path.join(self.basedir, "VERSION")) as fin:
            version = fin.read().rstrip()
        if version != expected_version:
            raise Exception("incorrect concat-fuse version, expected %s got %s" % (expected_version, version))

    def concat(self, files):
        """Send the file list to concat-fuse and return the filename of the virtual file"""

        # write the file list to concat-fuse
        files_serialized = "\0".join(files)
        with open(os.path.join(self.basedir, "from-file0/control"), "wb") as fout:
            fout.write(files_serialized)

        # return the virtual file name
        multifile_id = hashlib.sha1(files_serialized).hexdigest()
        virtual_file = os.path.join(self.basedir, "from-file0", multifile_id)
        return virtual_file

    def glob(self, globs):
        """Send the glob list to concat-fuse and return the filename of the virtual file"""

        # write the file list to concat-fuse
        globs_serialized = "\0".join(globs)
        with open(os.path.join(self.basedir, "from-glob0/control"), "wb") as fout:
            fout.write(globs_serialized)

        # return the virtual file name
        multifile_id = hashlib.sha1(globs_serialized).hexdigest()
        virtual_file = os.path.join(self.basedir, "from-glob0", multifile_id)
        return virtual_file

    def zip(self, zip):
        """Send the zip to concat-fuse and return the filename of the virtual file"""

        # write the zip filename list to concat-fuse
        with open(os.path.join(self.basedir, "from-zip/control"), "wb") as fout:
            fout.write(zip)

        # return the virtual file name
        multifile_id = hashlib.sha1(zip).hexdigest()
        virtual_file = os.path.join(self.basedir, "from-zip", multifile_id)
        return virtual_file


def splitlines0(data):
    lst = data.split("\0")
    if lst and lst[-1] == "":
        del lst[-1]
    return lst


def parse_args():
    parser = argparse.ArgumentParser(description="Create virtual concatenated files.",
                                     epilog="The name of the virtual file is returned on stdout.")

    # argparser does not allow freely mixing optional and positional
    # arguments, this limits the functionality of `cfconcat` somewhat, but
    # shouldn't be to critical: http://bugs.python.org/issue14191

    parser.add_argument('-v', '--version', action='store_true', help="Print version number")

    concat_fuse_group = parser.add_argument_group("concat-fuse options")
    concat_fuse_group.add_argument('-u', '--unmount', action='store_true', help="Unmount concat-fuse")
    concat_fuse_group.add_argument('-e', '--exe', metavar="EXE", type=str, help="Use EXE instead of concat-fuse")
    concat_fuse_group.add_argument('-m', '--mountpoint', metavar="MOUNTPOINT", type=str, default=None,
                                   help="Use MOUNTPOINT instead of default")

    general_group = parser.add_argument_group("general options")
    general_group.add_argument('--ext', metavar="EXT", type=str, help="Add an extension to the virtual filename")

    files_group = parser.add_argument_group("file list options", "Options for static file lists.")
    files_group.add_argument('-n', '--dry-run', action='store_true', help="Don't send file list to concat-fuse, print to stdout")
    files_group.add_argument('-k', '--keep', action='store_true', help="Continue even when files are missing")
    files_group.add_argument('FILES', nargs='*', help="Files to virtually concatenate")
    files_group.add_argument('--from-file', metavar='FILE', action='append', default=[], help="Read files from FILE, newline separated")
    files_group.add_argument('--from-file0', metavar='FILE', action='append', default=[], help="Read files from FILE, '\\0' separated")
    files_group.add_argument('-', dest="stdin", action='store_true', help="Read files from stdin")
    files_group.add_argument('-0', dest="stdin0", action='store_true', help="Read files from stdin, '\\0' separated")

    glob_group = parser.add_argument_group("glob options",
                                            ("Glob pattern act as dynamic file lists and the underlying virtual "
                                            "file can be dynamically updated when new files arrive."))
    glob_group.add_argument('-g', '--glob', metavar='GLOB', action='append', default=[], help="Use glob pattern to select files")

    zip_group = parser.add_argument_group("zip options",
                                            ("Read files from a .zip archive"))
    zip_group.add_argument('-z', '--zip', metavar='ARCHIVE', action='store', help="Read files from ARCHIVE")

    return parser.parse_args()


def make_virtual_filename(args):
    # generate the file list
    files = []

    if args.stdin:
        files.extend(sys.stdin.read().splitlines())
    elif args.stdin0:
        files.extend(splitlines0(sys.stdin.read()))

    for f in args.from_file:
        with open(f, "r") as fin:
            files.extend(fin.read().splitlines())

    for f in args.from_file0:
        with open(f, "rb") as fin:
            files.extend(splitlines0(fin.read()))

    files.extend(args.FILES)

    files = [os.path.abspath(f) for f in files]

    # generate glob list
    globs = []
    globs.extend(args.glob)
    globs = [os.path.abspath(g) for g in globs]

    # check if arguments are valid
    if globs and files :
        raise Exception("Can't mix globs and regular files")
    elif globs and args.zip:
        raise Exception("Can't mix globs and zip files")
    elif files and args.zip:
        raise Exception("Can't mix files and zip files")

    errors = False
    for f in files:
        if not os.path.exists(f):
            print("%s: No such file or directory" % f, file=sys.stderr)
            errors = True

    if errors and not args.keep:
        raise Exception("detected errors when scanning files")

    if files:
        if args.dry_run:
            for f in files:
                print(f)
        else:
            concat_fuse = ConcatFuse(args.mountpoint, args.exe)
            virtual_filename = concat_fuse.concat(files)
            return virtual_filename
    elif globs:
        if args.dry_run:
            for g in globs:
                print(g)
        else:
            concat_fuse = ConcatFuse(args.mountpoint, args.exe)
            virtual_filename = concat_fuse.glob(globs)
            return virtual_filename
    elif args.zip:
        zip_filename = os.path.abspath(args.zip)
        if not os.path.exists(zip_filename):
            raise Exception("%s: No such file or directory" % zip_filename)
        else:
            concat_fuse = ConcatFuse(args.mountpoint, args.exe)
            virtual_filename = concat_fuse.zip(zip_filename)
            return virtual_filename
    else:
        raise Exception("{}: fatal error: no input files provided" % sys.argv[0])


def main():
    args = parse_args()

    if args.version:
        print("cfconcat 0.3.2")
    elif args.unmount:
        ConcatFuse.unmount(args.mountpoint)
    else:
        try:
            virtual_filename = make_virtual_filename(args)
            if args.ext is not None:
                dir = tempfile.mkdtemp()
                link_name = os.path.join(dir, os.path.basename(virtual_filename) + args.ext)
                os.symlink(virtual_filename, link_name)
                print(link_name)
            else:
                print(virtual_filename)
        except Exception as err:
            print("error: {}".format(err), file=sys.stderr)
            sys.exit(1)


if __name__ == "__main__":
    main()


# EOF #
